### PR TITLE
feat!: multi tile storage destinations including fs

### PR DESCRIPTION
**Pull request by melancholiai**
| Question                | Answer                                                                          |
| ---------------- | -------------------------------------------------------------------------- |
| Bug fix         | ✖                                                                        |
| New feature     | ✔                                                                        |
| Breaking change | ✔                                                                        |
| Deprecations    | ✖                                                                        |
| Documentation   | ✔                                                                        |
| Tests added     | ✔                                                                        |
| Chore            | ✖                                                                       |

### CODE REVIEW

File: config/custom-environment-variables.json [hunk 1]
```diff
@@ -81,15 +81,10 @@
...
-      "s3ClientConfig": {
-        "endpoint": "S3_ENDPOINT",
-        "region": "S3_REGION",
-        "forcePathStyle": {
-          "__name": "S3_FORCE_PATH_STYLE",
-          "__format": "boolean"
-        }
+      "providers": {
+        "__name": "TILES_STORAGE_PROVIDERS",
+        "__format": "json"
```
Comment (syncush) [inline config/custom-environment-variables.json:86]: we can't validate its schema right?

File: src/index.ts [hunk 2]
```diff
@@ -29,11 +38,13 @@ void registerExternalValues()
...
     const server = livenessFactory(createServer(app));
 
-    shutdownHandler.addFunction(async () => {
-      return new Promise((resolve) => {
-        server.once('close', resolve);
-        server.close();
-      });
+    cleanupRegistry.register({
+      func: async () => {
+        return new Promise((resolve) => {
```
Comment (syncush) [inline src/index.ts:43]: shouldnt we use `server.drain` also ?

File: src/retiler/jobQueueProvider/pgBossJobQueue.ts [hunk 3]
```diff
@@ -15,25 +15,27 @@ export class PgBossJobQueueProvider implements JobQueueProvider {
   private runningJobs = 0;
 
   private readonly jobFinishedEmitter = new EventEmitter();
-  private readonly jobFinishedEventName = 'avi';
+  private readonly jobFinishedEventName = 'jobFinished';
```
Comment (syncush) [inline src/retiler/jobQueueProvider/pgBossJobQueue.ts:18]: unacceptable!

File: src/retiler/tilesStorageProvider/factory.ts [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+      if (!s3Client) {
+        s3Client = new S3Client(clientConfig);
+        s3ClientsMap.set(clientConfig.endpoint, s3Client);
+
+        cleanupRegistry.register({
+          func: async () => {
+            return new Promise((resolve) => {
+              (s3Client as S3Client).destroy();
+              resolve(undefined);
```
Comment (syncush) [inline src/retiler/tilesStorageProvider/factory.ts:33]: isn't it better to return `resolve(undefined);`?

File: src/retiler/tilesStorageProvider/fs.ts [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+
+    this.logger.debug({ msg: 'storing tile in fs', tile: baseTile, parent, baseStoragePath: this.baseStoragePath, key });
+
+    const storagePath = join(this.baseStoragePath, key);
+
+    try {
+      const dir = dirname(storagePath);
+      if (!existsSync(dir)) {
+        await mkdir(dir, { recursive: true });
+      }
```
Comment (syncush) [inline src/retiler/tilesStorageProvider/fs.ts:31]: should we make a dir on our own ? what about permissions and so on, I think the server should fail if path does not exists or is not writeable or readable. we should not change the state of the storage.

File: tests/configurations/integration/jest.config.js [hunk 6]
```diff
@@ -32,7 +32,7 @@ module.exports = {
       branches: 80,
       functions: 80,
       lines: 80,
-      statements: -10,
+      statements: -20,
```
Comment (syncush) [inline tests/configurations/integration/jest.config.js:35]: do we have a reason to lower the standard?

**Review by syncush — CHANGES_REQUESTED** [review]

File: config/custom-environment-variables.json (hunk 1 above)
Comment (melancholiai) [reply config/custom-environment-variables.json:86]: added ajv on the factory function

File: src/index.ts (hunk 2 above)
Comment (melancholiai) [reply src/index.ts:43]: no such function

File: src/retiler/jobQueueProvider/pgBossJobQueue.ts (hunk 3 above)
Comment (melancholiai) [reply src/retiler/jobQueueProvider/pgBossJobQueue.ts:18]: sorry ;(

File: src/retiler/tilesStorageProvider/factory.ts (hunk 4 above)
Comment (melancholiai) [reply src/retiler/tilesStorageProvider/factory.ts]: done.

File: src/retiler/tilesStorageProvider/fs.ts (hunk 5 above)
Comment (melancholiai) [reply src/retiler/tilesStorageProvider/fs.ts:31]: the tiles storage dir is set according to the fs storageProvider's `basePath` + `app.tilesStorage.layout.format` configuration value, the latter value changes from time to time.
I don't see a downside to attempt to create the dir, remember it also creates the `z/x/y` directories.
another manual step is another pain in the ass, and anyway if there are insufficient permissions to write the file the job will fail with a proper error.

File: tests/configurations/integration/jest.config.js (hunk 6 above)
Comment (melancholiai) [reply tests/configurations/integration/jest.config.js]: done, I've added a couple of tests to achieve the standard

**Review by syncush — APPROVED** [review]

**Pull request merged**